containers/src/Data/IntMap/Internal.hs

```diff
@@ -2266,37 +2266,37 @@ deleteFindMax = fromMaybe (error "deleteFindMax: empty map has no maximal elemen
 deleteFindMin :: IntMap a -> ((Key, a), IntMap a)
 deleteFindMin = fromMaybe (error "deleteFindMin: empty map has no minimal element") . minViewWithKey
 
+lookupMinSure :: IntMap a -> (Key, a)
+lookupMinSure (Tip k v)     = (k,v)
+lookupMinSure (Bin _ _ l _) = lookupMinSure l
+lookupMinSure Nil           = error "lookupMinSure Nil"
```

Comment:
We don't want to box the returned `Int` unnecessarily. I'd use this:

```suggestion
lookupMinSure :: IntMap a -> IntPair a
lookupMinSure (Tip k v)     = IntPair k v
lookupMinSure (Bin _ _ l _) = lookupMinSure l
lookupMinSure Nil           = error "lookupMinSure Nil"

data IntPair a = IntPair !Int a
```

Reply:
Speaking of which, this should probably be harmonized with `Map` by making `Data.Map.Internal.lookupMinSure` return a half-strict pair, so the strictness analyzer doesn't lose track of the fact that the key is in WHNF. (Of course, it's possible it already gets that right; but that should be verified.)

Reply:
> We don't want to box the returned Int unnecessarily.

I thought this is a good idea and was about to implement it, but then noticed that GHC is smart enough to make this unnecessary. The worker for `lookupMinSure` returns an unboxed int already. It has type `IntMap a -> (# Int#, a #)`.  
But, this is not the case for `lookup{LE,LT,GE,GT}`. I think we can improve these. I will take a shot at this separately.

> does wrapping the Just on outside still work/help now that GHC does a lot more fancy join point stuff?

I don't know much about join points, but I'll post the core output below so we can be satisfied that it looks good.

> this should probably be harmonized with Map by making Data.Map.Internal.lookupMinSure return a half-strict pair, so the strictness analyzer doesn't lose track of the fact that the key is in WHNF.

I don't know how to read this information, but I switched to a half-strict pair and found no difference in the core. I'll share it below for the record.

The core below is with GHC 9.6.3.

<details>
<summary>IntMap lookupMin (current)</summary>

```hs
-- RHS size: {terms: 23, types: 24, coercions: 0, joins: 0/0}
lookupMin :: forall a. IntMap a -> Maybe (Key, a)
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [70] 121 20}]
lookupMin
  = \ (@a_aQzr) (ds_dTMk :: IntMap a_aQzr) ->
      case ds_dTMk of {
        Bin bx_dW5u bx1_dW5v l_aOQ2 r_aOQ3 ->
          case GHCExts.<# bx1_dW5v 0# of {
            __DEFAULT -> unsafeFindMin @a_aQzr l_aOQ2;
            1# -> unsafeFindMin @a_aQzr r_aOQ3
          };
        Tip bx_dW5t v_aOQ0 ->
          GHC.Maybe.Just
            @(Key, a_aQzr) (ghc-prim:GHC.Types.I# bx_dW5t, v_aOQ0);
        Nil -> GHC.Maybe.Nothing @(Key, a_aQzr)
      }

Rec {
-- RHS size: {terms: 15, types: 22, coercions: 0, joins: 0/0}
unsafeFindMin [Occ=LoopBreaker]
  :: forall a. IntMap a -> Maybe (Key, a)
[GblId, Arity=1, Str=<1L>, Unf=OtherCon []]
unsafeFindMin
  = \ (@a_aQR0) (ds_dUaJ :: IntMap a_aQR0) ->
      case ds_dUaJ of {
        Bin bx_dW69 bx1_dW6a l_aOG2 ds1_dUcE ->
          unsafeFindMin @a_aQR0 l_aOG2;
        Tip bx_dW68 y_aOG1 ->
          GHC.Maybe.Just
            @(Key, a_aQR0) (ghc-prim:GHC.Types.I# bx_dW68, y_aOG1);
        Nil -> GHC.Maybe.Nothing @(Key, a_aQR0)
      }
end Rec }
```

Simple usage:
```hs
foo :: IntMap Int -> Int
foo = maybe 0 (\(k,v) -> k+v) . lookupMin

-- the worker compiles to core

-- RHS size: {terms: 55, types: 49, coercions: 0, joins: 0/0}
Main.$wfoo [InlPrag=NOINLINE]
  :: IntMap Int -> ghc-prim:GHC.Prim.Int#
[GblId[StrictWorker([!])], Arity=1, Str=<1L>, Unf=OtherCon []]
Main.$wfoo
  = \ (x_slv6 :: IntMap Int) ->
      case x_slv6 of {
        Bin bx_abUd bx1_abUe l_abUf r_abUg ->
          case ghc-prim:GHC.Prim.<# bx1_abUe 0# of {
            __DEFAULT ->
              case Data.IntMap.Internal.unsafeFindMin @Int l_abUf of {
                Nothing -> 0#;
                Just x29_abU8 ->
                  case x29_abU8 of { (k_a6lh, v2_a6li) ->
                  case k_a6lh of { ghc-prim:GHC.Types.I# x30_acEK ->
                  case v2_a6li of { ghc-prim:GHC.Types.I# y_acEN ->
                  ghc-prim:GHC.Prim.+# x30_acEK y_acEN
                  }
                  }
                  }
              };
            1# ->
              case Data.IntMap.Internal.unsafeFindMin @Int r_abUg of {
                Nothing -> 0#;
                Just x29_abU8 ->
                  case x29_abU8 of { (k_a6lh, v2_a6li) ->
                  case k_a6lh of { ghc-prim:GHC.Types.I# x30_acEK ->
                  case v2_a6li of { ghc-prim:GHC.Types.I# y_acEN ->
                  ghc-prim:GHC.Prim.+# x30_acEK y_acEN
                  }
                  }
                  }
              }
          };
        Tip bx_abUk v2_abUl ->
          case v2_abUl of { ghc-prim:GHC.Types.I# y_acEN ->
          ghc-prim:GHC.Prim.+# bx_abUk y_acEN
          };
        Nil -> 0#
      }
```

</details>

<details>
<summary>IntMap lookupMin (with this PR)</summary>

```hs
Rec {
-- RHS size: {terms: 14, types: 22, coercions: 0, joins: 0/0}
Data.IntMap.Internal.$wlookupMinSure [InlPrag=[2], Occ=LoopBreaker]
  :: forall {a}. IntMap a -> (# GHCExts.Int#, a #)
[GblId[StrictWorker([!])], Arity=1, Str=<1L>, Unf=OtherCon []]
Data.IntMap.Internal.$wlookupMinSure
  = \ (@a_sXtJ) (ds_sXtK :: IntMap a_sXtJ) ->
      case ds_sXtK of {
        Bin bx_dWca bx1_dWcb l_aOWN ds1_dTUy ->
          Data.IntMap.Internal.$wlookupMinSure @a_sXtJ l_aOWN;
        Tip bx_dWc9 v_aOWM -> (# bx_dWc9, v_aOWM #);
        Nil -> case lvl47_rZ02 of wild1_00 { }
      }
end Rec }

-- RHS size: {terms: 10, types: 13, coercions: 0, joins: 0/0}
lookupMinSure [InlPrag=[2]] :: forall a. IntMap a -> (Key, a)
[GblId,
 Arity=1,
 Str=<1L>,
 Cpr=1(1,),
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=1,unsat_ok=True,boring_ok=False)
         Tmpl= \ (@a_sXtJ) (ds_sXtK [Occ=Once1] :: IntMap a_sXtJ) ->
                 case Data.IntMap.Internal.$wlookupMinSure @a_sXtJ ds_sXtK of
                 { (# ww_sXVa [Occ=Once1], ww1_sXtP [Occ=Once1] #) ->
                 (ghc-prim:GHC.Types.I# ww_sXVa, ww1_sXtP)
                 }}]
lookupMinSure
  = \ (@a_sXtJ) (ds_sXtK :: IntMap a_sXtJ) ->
      case Data.IntMap.Internal.$wlookupMinSure @a_sXtJ ds_sXtK of
      { (# ww_sXVa, ww1_sXtP #) ->
      (ghc-prim:GHC.Types.I# ww_sXVa, ww1_sXtP)
      }

-- RHS size: {terms: 37, types: 48, coercions: 0, joins: 0/0}
lookupMin [InlPrag=INLINE (sat-args=1)]
  :: forall a. IntMap a -> Maybe (Key, a)
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=StableUser, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=1,unsat_ok=False,boring_ok=False)
         Tmpl= \ (@a_aQGE) (ds_dTUz [Occ=Once1!] :: IntMap a_aQGE) ->
                 case ds_dTUz of {
                   Bin _ [Occ=Dead] bx1_dWcd [Occ=Once1] l_aOWR [Occ=Once1]
                       r_aOWS [Occ=Once1] ->
                     case ghc-prim:GHC.Classes.ltInt
                            (ghc-prim:GHC.Types.I# bx1_dWcd) (ghc-prim:GHC.Types.I# 0#)
                     of {
                       False ->
                         case lookupMinSure @a_aQGE l_aOWR of vx_ie5p [Occ=Once1]
                         { (_ [Occ=Dead], _ [Occ=Dead]) ->
                         GHC.Maybe.Just @(Key, a_aQGE) vx_ie5p
                         };
                       True ->
                         case lookupMinSure @a_aQGE r_aOWS of vx_ie5p [Occ=Once1]
                         { (_ [Occ=Dead], _ [Occ=Dead]) ->
                         GHC.Maybe.Just @(Key, a_aQGE) vx_ie5p
                         }
                     };
                   Tip bx_dWce [Occ=Once1] v_aOWP [Occ=Once1] ->
                     GHC.Maybe.Just
                       @(Key, a_aQGE) (ghc-prim:GHC.Types.I# bx_dWce, v_aOWP);
                   Nil -> GHC.Maybe.Nothing @(Key, a_aQGE)
                 }}]
lookupMin
  = \ (@a_aQGE) (eta_B0 :: IntMap a_aQGE) ->
      case eta_B0 of {
        Bin bx_dWcc bx1_dWcd l_aOWR r_aOWS ->
          case GHCExts.<# bx1_dWcd 0# of {
            __DEFAULT ->
              case Data.IntMap.Internal.$wlookupMinSure @a_aQGE l_aOWR of
              { (# ww_sXVa, ww1_sXtP #) ->
              GHC.Maybe.Just
                @(Key, a_aQGE) (ghc-prim:GHC.Types.I# ww_sXVa, ww1_sXtP)
              };
            1# ->
              case Data.IntMap.Internal.$wlookupMinSure @a_aQGE r_aOWS of
              { (# ww_sXVa, ww1_sXtP #) ->
              GHC.Maybe.Just
                @(Key, a_aQGE) (ghc-prim:GHC.Types.I# ww_sXVa, ww1_sXtP)
              }
          };
        Tip bx_dWce v_aOWP ->
          GHC.Maybe.Just
            @(Key, a_aQGE) (ghc-prim:GHC.Types.I# bx_dWce, v_aOWP);
        Nil -> GHC.Maybe.Nothing @(Key, a_aQGE)
      }
```

Simple usage:
```hs
foo :: IntMap Int -> Int
foo = maybe 0 (\(k,v) -> k+v) . lookupMin

-- the worker compiles to core

-- RHS size: {terms: 39, types: 35, coercions: 0, joins: 0/0}
Main.$wfoo [InlPrag=NOINLINE]
  :: IntMap Int -> ghc-prim:GHC.Prim.Int#
[GblId[StrictWorker([!])], Arity=1, Str=<1L>, Unf=OtherCon []]
Main.$wfoo
  = \ (x_sjZs :: IntMap Int) ->
      case x_sjZs of {
        Bin bx_ib6U bx1_ib6V l_ib6W r_ib6X ->
          case ghc-prim:GHC.Prim.<# bx1_ib6V 0# of {
            __DEFAULT ->
              case Data.IntMap.Internal.$wlookupMinSure @Int l_ib6W of
              { (# ww_ij2b, ww1_ij2c #) ->
              case ww1_ij2c of { ghc-prim:GHC.Types.I# y_aaE9 ->
              ghc-prim:GHC.Prim.+# ww_ij2b y_aaE9
              }
              };
            1# ->
              case Data.IntMap.Internal.$wlookupMinSure @Int r_ib6X of
              { (# ww_ij2b, ww1_ij2c #) ->
              case ww1_ij2c of { ghc-prim:GHC.Types.I# y_aaE9 ->
              ghc-prim:GHC.Prim.+# ww_ij2b y_aaE9
              }
              }
          };
        Tip bx_ib7b v2_ib7c ->
          case v2_ib7c of { ghc-prim:GHC.Types.I# y_aaE9 ->
          ghc-prim:GHC.Prim.+# bx_ib7b y_aaE9
          };
        Nil -> 0#
      }
```

</details>

<details>
<summary>Map lookupMin</summary>

```hs
Rec {
-- RHS size: {terms: 16, types: 25, coercions: 0, joins: 0/0}
Data.Map.Internal.$wlookupMinSure [InlPrag=[2], Occ=LoopBreaker]
  :: forall {k} {a}. k -> a -> Map k a -> (# k, a #)
[GblId[StrictWorker([~, ~, !])],
 Arity=3,
 Str=<L><L><1L>,
 Unf=OtherCon []]
Data.Map.Internal.$wlookupMinSure
  = \ (@k_s3BIC)
      (@a_s3BID)
      (k1_s3BIE :: k_s3BIC)
      (a1_s3BIF :: a_s3BID)
      (ds_s3BIG :: Map k_s3BIC a_s3BID) ->
      case ds_s3BIG of {
        Bin bx_d3AF5 k2_a3mSH a2_a3mSI l_a3mSJ ds1_d3sTN ->
          Data.Map.Internal.$wlookupMinSure
            @k_s3BIC @a_s3BID k2_a3mSH a2_a3mSI l_a3mSJ;
        Tip -> (# k1_s3BIE, a1_s3BIF #)
      }
end Rec }

-- RHS size: {terms: 14, types: 18, coercions: 0, joins: 0/0}
lookupMinSure [InlPrag=[2]]
  :: forall k a. k -> a -> Map k a -> (k, a)
[GblId,
 Arity=3,
 Str=<L><L><1L>,
 Cpr=1,
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=3,unsat_ok=True,boring_ok=False)
         Tmpl= \ (@k_s3BIC)
                 (@a_s3BID)
                 (k1_s3BIE [Occ=Once1] :: k_s3BIC)
                 (a1_s3BIF [Occ=Once1] :: a_s3BID)
                 (ds_s3BIG [Occ=Once1] :: Map k_s3BIC a_s3BID) ->
                 case Data.Map.Internal.$wlookupMinSure
                        @k_s3BIC @a_s3BID k1_s3BIE a1_s3BIF ds_s3BIG
                 of
                 { (# ww_s3CeR [Occ=Once1], ww1_s3CeS [Occ=Once1] #) ->
                 (ww_s3CeR, ww1_s3CeS)
                 }}]
lookupMinSure
  = \ (@k_s3BIC)
      (@a_s3BID)
      (k1_s3BIE :: k_s3BIC)
      (a1_s3BIF :: a_s3BID)
      (ds_s3BIG :: Map k_s3BIC a_s3BID) ->
      case Data.Map.Internal.$wlookupMinSure
             @k_s3BIC @a_s3BID k1_s3BIE a1_s3BIF ds_s3BIG
      of
      { (# ww_s3CeR, ww1_s3CeS #) ->
      (ww_s3CeR, ww1_s3CeS)
      }

-- RHS size: {terms: 18, types: 34, coercions: 0, joins: 0/0}
lookupMin :: forall k a. Map k a -> Maybe (k, a)
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [30] 90 20}]
lookupMin
  = \ (@k_a3p4C) (@a_a3p4D) (ds_d3sTO :: Map k_a3p4C a_a3p4D) ->
      case ds_d3sTO of {
        Bin bx_d3AF6 k1_a3mSK x_a3mSL l_a3mSM ds1_d3sWe ->
          case Data.Map.Internal.$wlookupMinSure
                 @k_a3p4C @a_a3p4D k1_a3mSK x_a3mSL l_a3mSM
          of
          { (# ww_s3CeR, ww1_s3CeS #) ->
          GHC.Maybe.Just @(k_a3p4C, a_a3p4D) (ww_s3CeR, ww1_s3CeS)
          };
        Tip -> GHC.Maybe.Nothing @(k_a3p4C, a_a3p4D)
      }
```

</details>

Reply:
Ah, I guess GHC's newish nested product stuff actually works! I'm a bit surprised the `Int` operations weren't inlined in the unfolding; I wonder why that is. When you tried changing the `Map` version, did the worker(s?) end up with the same strictness signature and CPR info?

Reply:
>  I'm a bit surprised the Int operations weren't inlined in the unfolding

Which ones do you mean?

> When you tried changing the Map version, did the worker(s?) end up with the same strictness signature and CPR info?

Ah there is small diff, sorry I overlooked this. I made this change
```diff
-lookupMinSure :: k -> a -> Map k a -> (k, a)
-lookupMinSure k a Tip = (k, a)
+data KV k a = KV !k a
+
+lookupMinSure :: k -> a -> Map k a -> KV k a
+lookupMinSure k a Tip = KV k a
 lookupMinSure _ _ (Bin _ k a l _) = lookupMinSure k a l
```
which caused this change in the worker
```diff
 Rec {
--- RHS size: {terms: 16, types: 25, coercions: 0, joins: 0/0}
+-- RHS size: {terms: 19, types: 26, coercions: 0, joins: 0/0}
 Data.Map.Internal.$wlookupMinSure [InlPrag=[2], Occ=LoopBreaker]
   :: forall {k} {a}. k -> a -> Map k a -> (# k, a #)
 [GblId[StrictWorker([~, ~, !])],
  Arity=3,
- Str=<L><L><1L>,
+ Str=<ML><L><1L>,
  Unf=OtherCon []]
 Data.Map.Internal.$wlookupMinSure
   = \ (@k) (@a) (k1 :: k) (a1 :: a) (ds :: Map k a) ->
       case ds of {
         Bin bx k2 a2 l ds1 ->
           Data.Map.Internal.$wlookupMinSure @k @a k2 a2 l;
-        Tip -> (# k1, a1 #)
+        Tip -> case k1 of conrep { __DEFAULT -> (# conrep, a1 #) }
       }
 end Rec }
 ```
 Not sure if this is beneficial.

Reply:
I find demand signatures pretty mysterious too, but you can look at a use case like

```haskell
toKV :: (k, a) -> KV k a

test = fmap toKV . lookupMin
```

I expect the current implementation to stick a thunk in the `Just` and I expect my proposed one not to.

Reply:
I see what you mean.

```hs
test :: Map Int Int -> Maybe (KV Int Int)
test = fmap toKV . lookupMin
```

<details>
<summary>Current</summary>

```hs
-- RHS size: {terms: 16, types: 32, coercions: 0, joins: 0/0}
test :: Map Int Int -> Maybe (KV Int Int)
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [30] 110 20}]
test
  = \ (x_in9W :: Map Int Int) ->
      case x_in9W of {
        Bin bx_dmQo k_a6s7 x1_a6s8 l_a6s9 ds_df6l ->
          case Data.Map.Internal.$wlookupMinSure
                 @Int @Int k_a6s7 x1_a6s8 l_a6s9
          of
          { (# ww_soRW, ww1_soRX #) ->
          GHC.Maybe.Just
            @(KV Int Int) (Data.Map.Internal.$WKV @Int @Int ww_soRW ww1_soRX)
          };
        Tip -> GHC.Maybe.Nothing @(KV Int Int)
      }
```

And `$WKV` is

```hs
-- RHS size: {terms: 10, types: 7, coercions: 0, joins: 0/0}
Data.Map.Internal.$WKV [InlPrag=INLINE[final] CONLIKE]
  :: forall k a. k %1 -> a %1 -> KV k a
[GblId[DataConWrapper],
 Arity=2,
 Str=<SL><L>,
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (@k_a69k)
                 (@a_a69l)
                 (conrep_a7Fm [Occ=Once1] :: k_a69k)
                 (conrep1_a7Fn [Occ=Once1] :: a_a69l) ->
                 case conrep_a7Fm of conrep2_a7Fm [Occ=Once1] { __DEFAULT ->
                 Data.Map.Internal.KV @k_a69k @a_a69l conrep2_a7Fm conrep1_a7Fn
                 }}]
Data.Map.Internal.$WKV
  = \ (@k_a69k)
      (@a_a69l)
      (conrep_a7Fm [Occ=Once1] :: k_a69k)
      (conrep1_a7Fn [Occ=Once1] :: a_a69l) ->
      case conrep_a7Fm of conrep2_a7Fm [Occ=Once1] { __DEFAULT ->
      Data.Map.Internal.KV @k_a69k @a_a69l conrep2_a7Fm conrep1_a7Fn
      }
```
</details>

<details>
<summary>If lookupSure returns KV</summary>

```hs
-- RHS size: {terms: 16, types: 32, coercions: 0, joins: 0/0}
test :: Map Int Int -> Maybe (KV Int Int)
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [30] 90 20}]
test
  = \ (x_inar :: Map Int Int) ->
      case x_inar of {
        Bin bx_dmQN k_a6s7 x1_a6s8 l_a6s9 ds_df6K ->
          case Data.Map.Internal.$wlookupMinSure
                 @Int @Int k_a6s7 x1_a6s8 l_a6s9
          of
          { (# ww_soSo, ww1_soSp #) ->
          GHC.Maybe.Just
            @(KV Int Int) (Data.Map.Internal.KV @Int @Int ww_soSo ww1_soSp)
          };
        Tip -> GHC.Maybe.Nothing @(KV Int Int)
      }
```
</details>

How important do you think this is? The user can use `(<$!>)` instead of `fmap` with the current `lookupMin` to avoid the thunk. If `lookupSure` returns `KV`, the `case k1 of conrep ...` is moving from `test` to `lookupSure`.  
Also, if we do this, should `Map.Strict` reimplement this with a `data KV' k v = KV !k !a`?

Reply:
> How important do you think this is?

I think it's ... semi-important. I'd like to get it right.

> The user can use `(<$!>)` instead of `fmap` with the current `lookupMin` to avoid the thunk.

I think that will end up (cheaply) forcing something already in WHNF.


> If `lookupMinSure` returns `KV`, the `case k1 of conrep ...` is moving from `test` to `lookupMinSure`.

Huh? It shouldn't ... that should just get erased altogether, because the key is in WHNF in the `Map`.

> Also, if we do this, should `Map.Strict` reimplement this with a `data KV' k a = KV' !k !a`?

No, we don't want lookups to force values.

Reply:
> I think it's ... semi-important. I'd like to get it right.

👍 

> No, we don't want lookups to force values.

👍 

> I think that will end up (cheaply) forcing something already in WHNF.

It is forced in `lookupMinSure` anyway, but more on that below. If we can avoid forcing it that would be a win.

> Huh? It shouldn't ... that should just get erased altogether, because the key is in WHNF in the Map.

You can see it's not erased in the diff I posted (https://github.com/haskell/containers/pull/976#discussion_r1451831773). And it's understandable why, since `lookupMinSure` can be called with a non-WHNF `k` from outside, though for the recursive calls this will not be the case.

I don't know if we can write this in a way which makes GHC understand that `lookupMinSure` is never called with a non-WHNF `k`, short of making `lookupMinSure` a function local to `lookupMin`.
```hs
lookupMin = ...
  where
    lookupMinSure = ...
```

If we do this and `lookupMin` is `INLINE`, we would be inlining more than necessary. If we omit `INLINE`, I saw that it works ideally (`lookupMinSure` becomes a top-level worker and the `k` is not forced anywhere), but now we don't know for sure whether `lookupMin` will inline. At least this is what I understand. Do you prefer one of these? Or is there any other way we can get the ideal result?

Reply:
We definitely don't want to inline everything.... I guess if we just make `lookupMinSure` strict in the passed key then we force unnecessarily? An overkill approach might be

```haskell
data UBox :: Type -> TYPE UnliftedRep where
  UBox :: !a -> UBox a

lookupMinSure :: UBox k -> Map k a -> KV k a
```

This explicitly says that the argument to `lookupMinSure` must *already* be in WHNF when the function is called, which is exactly what we're trying to say. I would love to have a nicer way to say that. It would be nice to see how that behaves anyway, but it's *probably* not worth dragging in that much machinery.

In terms of judgement calls, I'd much rather risk an unnecessary (double) force than an unnecessary thunk allocation; the former is way cheaper.

Reply:
Correction: I was mistaken, even if `lookupMinSure` is local, there is still a `case k1 of conrep`. Sigh.

> I guess if we just make lookupMinSure strict in the passed key then we force unnecessarily?

With `lookupMinSure !k a = ...`, the case moves to the top, which seems worse to me since we're forcing on every iteration.

<details>
<summary>Core</summary>

```hs
Rec {
-- RHS size: {terms: 19, types: 26, coercions: 0, joins: 0/0}
Data.Map.Internal.$wlookupMinSure [InlPrag=[2], Occ=LoopBreaker]
  :: forall {k} {a}. k -> a -> Map k a -> (# k, a #)
[GblId[StrictWorker([!, ~, !])],
 Arity=3,
 Str=<1L><L><1L>,
 Unf=OtherCon []]
Data.Map.Internal.$wlookupMinSure
  = \ (@k_somT)
      (@a_somU)
      (k1_somV :: k_somT)
      (a1_somW :: a_somU)
      (ds_somX :: Map k_somT a_somU) ->
      case k1_somV of k2_X0 { __DEFAULT ->
      case ds_somX of {
        Bin ipv_snah ipv1_snai ipv2_snaj ipv3_snak ipv4_snal ->
          Data.Map.Internal.$wlookupMinSure
            @k_somT @a_somU ipv1_snai ipv2_snaj ipv3_snak;
        Tip -> (# k2_X0, a1_somW #)
      }
      }
end Rec }
```
</details>

I tried your `UBox` version
```hs
lookupMinSure (UBox k) a Tip = KV k a
lookupMinSure _ _ (Bin _ k a l _) = lookupMinSure (UBox k) a l
```

but it leads to the same core as the strict version above.

> In terms of judgement calls, I'd much rather risk an unnecessary (double) force than an unnecessary thunk allocation; the former is way cheaper.

Understood, I'll update the PR for Map then.

Reply:
I'm very curious why the `UBox` version would ever force the key. We might want to open a GHC ticket for that.

Reply:
Cool, I'll open an issue.

Reply:
Opened [GHC#24340](https://gitlab.haskell.org/ghc/ghc/-/issues/24340).